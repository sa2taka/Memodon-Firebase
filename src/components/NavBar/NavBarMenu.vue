<template>
  <div>
    <v-app-bar-nav-icon
      @click.stop="drawer = !drawer"
      v-if="isSmartphoneWidth"
    ></v-app-bar-nav-icon>
    <v-navigation-drawer
      v-model="drawer"
      absolute
      right
      temporary
      height="100vh"
      :color="isDark ? 'menu-grey' : 'white'"
      v-if="isSmartphoneWidth"
    >
      <v-list-item>
        <v-list-item-content>
          <v-list-item-title class="title">Memodon</v-list-item-title>
          <v-list-item-subtitle>Memodon is your note.</v-list-item-subtitle>
        </v-list-item-content>
      </v-list-item>

      <v-divider></v-divider>

      <v-list dense nav>
        <slot></slot>
      </v-list>
      <v-divider></v-divider>
      <common-settings
        :class="[isDark ? 'menu-grey' : 'white']"
      ></common-settings>
    </v-navigation-drawer>
    <v-menu
      v-if="!isSmartphoneWidth"
      bottom
      left
      :close-on-content-click="false"
      transition="slide-y-transition"
      offset-y
      :content-class="isDark ? 'menu-grey' : 'white'"
      :value="menu"
      @input="input($event)"
    >
      <template v-slot:activator="{ on }">
        <v-btn icon v-on="on" @click="drawer = !drawer">
          <v-icon>fa-bars</v-icon>
        </v-btn>
      </template>

      <v-list dense nav :color="isDark ? 'menu-grey' : 'white'">
        <slot></slot>
      </v-list>
      <v-divider></v-divider>
      <common-settings
        :class="[isDark ? 'menu-grey' : 'white']"
      ></common-settings>
    </v-menu>
  </div>
</template>

<script lang="ts">
import { Component, Watch, Emit, Prop, Vue } from 'vue-property-decorator';
import theme from '@/store/modules/theme';
import { MinSmallWidth } from '@/libs/globalConstVariables';
import CommonSettings from '@/components/NavBar/CommonSettings.vue';

@Component({
  components: {
    CommonSettings,
  },
})
export default class NavBarMenu extends Vue {
  @Prop({ required: true })
  public menu: boolean | undefined;

  private isDark: boolean = false;
  private isSmartphoneWidth: boolean = false;
  private BoundaryWidth = MinSmallWidth;
  private drawer: boolean = false;
  public created() {
    this.setTheme();
    this.subscribeTheme();
    this.detectDevice();
    this.subscribeResizeWindow();
  }

  private detectDevice() {
    if (window.innerWidth > this.BoundaryWidth) {
      this.isSmartphoneWidth = false;
    } else {
      this.isSmartphoneWidth = true;
    }
  }

  private setTheme() {
    this.isDark = theme.theme === 'dark';
  }

  private subscribeTheme() {
    this.$store.subscribe((mutation, state) => {
      if (mutation.type === 'theme/setTheme') {
        this.isDark = mutation.payload === 'dark';
      }
    });
  }

  private subscribeResizeWindow() {
    window.addEventListener('resize', () => {
      this.detectDevice();
    });
  }
  @Emit()
  public input(value: boolean) {}
}
</script>

<style lang="scss" scoped>
.menu-grey {
  background: #484848 !important;
}
</style>
