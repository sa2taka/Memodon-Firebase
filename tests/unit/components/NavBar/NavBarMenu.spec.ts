import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import Vuex from 'vuex';
import NavBarMenu from '@/components/NavBar/NavBarMenu.vue';

const localVue = createLocalVue();

localVue.use(Vuetify);

describe('Component', () => {
  let wrapper: any;

  beforeEach(() => {
    const store = new Vuex.Store({});
    wrapper = shallowMount(NavBarMenu, {
      store,
      localVue,
      vuetify: new Vuetify(),
    });
  });

  test('is a Vue instance', () => {
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('v-menu-stub').exists()).toBe(true);
    expect(wrapper.find('v-menu-stub').attributes().contentclass).toBe('white');
  });

  test('is dark mode', () => {
    wrapper.vm.isDark = true;
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('v-menu-stub').attributes().contentclass).toBe(
      'menu-grey'
    );
  });

  test('is smartphone mode', () => {
    wrapper.vm.isSmartphoneWidth = true;
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('v-navigation-drawer-stub').exists()).toBe(true);
  });
});
