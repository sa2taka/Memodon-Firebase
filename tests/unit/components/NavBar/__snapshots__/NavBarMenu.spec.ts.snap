// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is a Vue instance 1`] = `
<div>
  <!---->
   
  <!---->
   
  <v-menu-stub
    bottom="true"
    closedelay="0"
    closeonclick="true"
    contentclass="white"
    left="true"
    maxheight="auto"
    maxwidth="auto"
    nudgebottom="0"
    nudgeleft="0"
    nudgeright="0"
    nudgetop="0"
    nudgewidth="0"
    offsety="true"
    opendelay="0"
    openonclick="true"
    origin="top left"
    transition="slide-y-transition"
  >
     
    <v-list-stub
      color="white"
      dense="true"
      nav="true"
      tag="div"
      tile="true"
    />
  </v-menu-stub>
</div>
`;

exports[`Component is dark mode 1`] = `
<div>
  <!---->
   
  <!---->
   
  <v-menu-stub
    bottom="true"
    closedelay="0"
    closeonclick="true"
    contentclass="menu-grey"
    left="true"
    maxheight="auto"
    maxwidth="auto"
    nudgebottom="0"
    nudgeleft="0"
    nudgeright="0"
    nudgetop="0"
    nudgewidth="0"
    offsety="true"
    opendelay="0"
    openonclick="true"
    origin="top left"
    transition="slide-y-transition"
  >
     
    <v-list-stub
      color="menu-grey"
      dense="true"
      nav="true"
      tag="div"
      tile="true"
    />
  </v-menu-stub>
</div>
`;

exports[`Component is smartphone mode 1`] = `
<div>
  <v-app-bar-nav-icon-stub />
   
  <v-navigation-drawer-stub
    absolute="true"
    color="white"
    height="100vh"
    minivariantwidth="80"
    mobilebreakpoint="1264"
    right="true"
    src=""
    tag="aside"
    temporary="true"
    width="256"
  >
    <v-list-item-stub
      activeclass=""
      tag="div"
    >
      <v-list-item-content-stub>
        <v-list-item-title-stub
          class="title"
        >
          Memodon
        </v-list-item-title-stub>
         
        <v-list-item-subtitle-stub>
          Memodon is your note.
        </v-list-item-subtitle-stub>
      </v-list-item-content-stub>
    </v-list-item-stub>
     
    <v-divider-stub />
     
    <v-list-stub
      dense="true"
      nav="true"
      tag="div"
      tile="true"
    />
  </v-navigation-drawer-stub>
   
  <!---->
</div>
`;
